### ISSUE TITLE
String Manipulation Functions

### ISSUE THREAD

**Issue by Undistraction** [char0n/ramda-adjunct#391]
Ramda already has:

- [**toUpper**](http://ramdajs.com/docs/#toUpper)
- [**toLower**](http://ramdajs.com/docs/#toLower)
- [**trim**](http://ramdajs.com/docs/#trim)
- [**split**](http://ramdajs.com/docs/#split)

Under discussion:

- [Capitalize](https://github.com/char0n/ramda-adjunct/issues/389) `Alpha beta`

Proposed:

1. Case

- **toSnakecase**  `alpha_beta`
- **toKebabCase** / **dasherize** `alpha-beta`
- **toCamelCase** `alphaBeta`
- **toUpperCamelCase** / **to PascalCase** `AlphaBeta`
- **toTitleCase** `Alpha Beta`
- **toStartCase** Ref: [lodash.pad](https://lodash.com/docs/4.17.5#startCase)

2. Formatting

- **wrap()** `wrap(`'`)('alpha')` ->  `'alpha'` / `wrap('[', ']')('alpha')` ->  `[alpha]`
- **pad()** Ref: [lodash.pad](https://lodash.com/docs/4.17.5#pad)
- **padLeft()**
- **padRight()**

**Comment by guillaumearm** [comment]
Love `wrap` for creating pipelines ;-)

But is it a good idea to name it `wrap` ? maybe wrapString ?

for the others, names looks good to me.

**Comment by Undistraction** [comment]
[Link](https://github.com/Undistraction/folktale-validations/blob/master/src/utils/formatting.js) to some of my own formatting utils for reference.

**Comment by Undistraction** [comment]
Ref: [Rails string utils](https://apidock.com/rails/ActionView/Helpers/TextHelper)

**Comment by char0n** [comment]
@Undistraction dedent is a nice one too https://www.npmjs.com/package/dedent

**Comment by char0n** [comment]
Regarding toUpperCamelCase / toPascalCase

toPascalCase should be primary name and tuUpperCamelCase @alias.

**Comment by char0n** [comment]
oStartCase Ref: lodash.pad -> oStartCase Ref: lodash.startCase

**Comment by char0n** [comment]
toTitleCase with capitalize @alias

**Comment by Undistraction** [comment]
> dedent is a nice one too https://www.npmjs.com/package/dedent

Added

> Regarding toUpperCamelCase / toPascalCase

How about `toMacroCase` / `toScreamingSnakeCase`?

> oStartCase Ref: lodash.pad -> oStartCase Ref: lodash.startCase

Done

> toTitleCase with capitalize @alias

These are two different functions - `capitalise` capitalises only the first letter of the first word `toTitleCase` capitalises the first letter of every word.

**Comment by char0n** [comment]
> How about toMacroCase / toScreamingSnakeCase?

Would go for toMacroCase as primary and toScreamingSnakeCase as @alias

> These are two different functions - capitalise capitalises only the first letter of the first word toTitleCase capitalises the first letter of every word.

Ok does toCapitalCase and toTitleCase make sense ?

**Comment by Undistraction** [comment]
> Ok does toCapitalCase and toTitleCase make sense ?

I think they have different intents - the `to...Case` are transforming across the whole string, but `capitalize` will only ever change the first char.

**Comment by char0n** [comment]
Yeah agree `capitalize` is for sure better.

For kebab case and other cases we will need these utils to sanitize the input string (should be part of src/internal):

```js
// canonicalDecompose :: String -> String
export const canonicalDecompose = invoker(1, 'normalize')('NFD');

// removes combining characters from decomposed form
// removeAccents :: String -> String
export const removeAccents = replace(/[\u0300-\u036f]/g, '');
```

**Comment by Undistraction** [comment]
@char0n Lodash exposes removal of accents as [deburr](https://lodash.com/docs/4.17.5#deburr)

**Comment by char0n** [comment]
Cool so we'll combine these two into deburr

**Comment by Undistraction** [comment]
Question: what input are we accepting for all the `to...Case` functions. Should they only work on a spaced string - `alpha beta` or should they work on each other - for example should `toCamelCase` work on `alpha-beta` or `alpha.beta`?

**Comment by char0n** [comment]
I would expect them to work on each other. Lodash seems to agree.

**Comment by GingerPlusPlus** [comment]
`stripLinebreaks` -- `trim` is often used for that, it strips all whitespace from beggining and end of the string.

What `stripWhitespace` would be supposed to do?

`wrap` should have fixed arity of 3. And it wouldn't have to be limited to strings :)

**Comment by Undistraction** [comment]
@char0n You might be interested in a little lib I just published: https://github.com/Undistraction/dedent-preserving-indents. It works much like dedent, but preserves all indentation past the line with the least whitespace. Might be nice to expose as a utility through RA.

**Comment by char0n** [comment]
@Undistraction yeah why not, but it will have to reintegrated and not used as dependency.

**Comment by GingerPlusPlus** [comment]
> You will probably have noticed that the first line is empty in the examples above. If you don't put the first line of chars on a new line, the function will have no effect as there will be no leading whitespace for that line, meaning no leading whitespace will be removed from the other lines. Always put the first line on a newline.

@Undistraction how about calling `.trim()` on a string first? That'd resolve it.

**Comment by Undistraction** [comment]
@gingerplusplus I think you've failed to understand the purpose of this function.

**Comment by GingerPlusPlus** [comment]
I do understand `dedent-js`, I use it in one of my projects. Unsure what the difference from it gives you.

**Comment by Undistraction** [comment]
> but it will have to reintegrated and not used as dependency

@char0n Yep. I'll open an issue for it and we can look at it somewhere down the line.

**Comment by damiangreen** [comment]
Also `upperFirst`, `truncate` would be useful
